src/crypto/random.c

```diff
@@ -146,3 +146,20 @@ void generate_random_bytes_not_thread_safe(size_t n, void *result) {
     }
   }
 }
+
+void add_extra_entropy(const void *ptr, size_t bytes)
+{
+  char hash[HASH_SIZE];
+  int i;
+
+  while (bytes > 0)
+  {
+    const size_t round_bytes = bytes > HASH_SIZE ? HASH_SIZE : bytes;
+    cn_fast_hash(ptr, round_bytes, hash);
```

Comment:
Nothing in particular. It seemed like a safe way to add entropy. If you think another way is better, I'll change it to that (when I'm back).

Reply:
Another option is to use user-provided entropy bytes as-is without any additional hashing or permutations.
Would have to add minimal size restriction `bytes >= HASH_SIZE`.

Reply:
> Another option is to use user-provided entropy bytes as-is without any additional hashing or permutations.
> Would have to add minimal size restriction bytes >= HASH_SIZE.

I doubt we would want to do this. This would not be adding entropy, but instead would be replacing the entropy from `/dev/urandom` entirely in most cases (if I understand your suggestion correctly). The exceptional case is when random bytes are retrieved before calling this function - its then absorbing bytes into the state in a slightly different way than outlined in the Keccak PRNG paper.

Following the Keccak paper is fairly easy: put  `hash_permutation` at the top of the `while` loop, and XOR `min(bytes, HASH_DATA_AREA)` into the first 136-bytes of the state. There are a couple of other ways to do it that would require changes to the other functions; the aforementioned is the least amount of code AFAIK.

Reply:
Meant `ROL + XOR`-ing the state with user-provided bytes. Thinking about 
`entropy(buf) > entropy(cn_fast_hash(buf)), buf is 32 bytes`
case due to some yet unknown bug in `cn_fast_hash`.

Reply:
> Meant ROL + XOR-ing the state with user-provided bytes.

That does not seem preferrable over the more researched Keccak permutation function.

This will probably work if entropy is added once immediately after initializing the state, but is not desireable in any other case. The last 64-bytes of the state are the secret portion, and shall never be returned/leaked for any reason. So that design would be more fragile - controlling how this function executes would have to be tightly controlled (which is difficult since so many places are calling for random bytes).

> Thinking about `entropy(buf) > entropy(cn_fast_hash(buf)), buf is 32 bytes`
case due to some yet unknown bug in `cn_fast_hash`.

The _actual_ entropy provided by the user is not known to exceed the output of the hash, but I'm not sure what that leaves this analysis. The permutation function state is 200-bytes...

Reply:
> The actual entropy provided by the user is not known to exceed the output of the hash

For 32 bytes case, input entropy surely *might* exceed output entropy, based on the hashing function implementation (for example, due to some yet unknown bug in `cn_fast_hash`).

```c++
void cn_fast_hash(const void *data, std::size_t length, hash &hash) {
  ...
  hash.data[0] = 0;
}
```

Reply:
`cn_fast_hash` is using Keccak permutation function at its core, so the entire PRNG could be busted in that case.

And, `cn_fast_hash` (which truncates the output from Keccak) doesn't need to be called if following the paper directly - only XOR and `hash_permutation` (with 200-byte output) needs to be used.

Reply:
Looking at the Sponge construction PDF, it's not really clear to me what the details of the algorithm are. I assume you're refering to the paragraph that starts with "If the request is feed". If so, and if doing the permutation first is just the change you think is needed, I'll do that now. As for hash, it's basically me thinking hashing stuff is always a good way to remove a degree of freedom that's not wanted. Since the paper says we simply append the data, I'll remove it too.

Reply:
I pushed an update with what I think you asked.